pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,200 @@
+	// machineAPIChecker determines if the precondition for this controller is met,
+	// this controller can be run only on a cluster that exposes a functional Machine API
+	machineAPIChecker     ceohelpers.MachineAPIChecker
+	masterMachineSelector labels.Selector
+}
+
+// NewMachineDeletionHooksController only reconciles machine hooks for the voting members
+// primarily to avoid an inherent race between this controller and the one that promotes learners
+// the following scenario might happen:
+//  1. A new machine joins the cluster, initially as a learner
+//  2. This controller adds a hook for the new machine
+//  3. A user deletes the machine
+//  4. The other controller misses 3 and decides to promote to the voting member, since this is a network call it takes arbitrary long to finish
+//  5. This controller removes the hook because the machine was deleted and the etcd member wasn't promoted to a voting member
+//  6. The network call from 4 finishes and the member is promoted
+//  7. At some point, the machine is removed from the etcd cluster, and we lose a member in an uncontrolled way
+//
+// note:
+//  The Machine Deletion Hook this controller reconciles is a mechanism within the Machine API that allow this operator to hold up removal of a machine
+//  until a replacement member has been promoted to a voting member.
+func NewMachineDeletionHooksController(
+	operatorClient operatorv1helpers.OperatorClient,
+	machineClient machinev1beta1client.MachineInterface,
+	machineAPIChecker ceohelpers.MachineAPIChecker,
+	masterMachineSelector labels.Selector,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &machineDeletionHooksController{
+		machineClient:                     machineClient,
+		machineAPIChecker:                 machineAPIChecker,
+		masterMachineSelector:             masterMachineSelector,
+		configMapListerForTargetNamespace: kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Lister().ConfigMaps(operatorclient.TargetNamespace),
+	}
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(
+			kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Informer(),
+		).
+		ResyncEvery(7*time.Minute). // an arbitrary number, not too high not too low
+		ToController("MachineDeletionHooksController", eventRecorder.WithComponentSuffix("cluster-machine-deletion-hooks-controller"))
+}
+
+func (c *machineDeletionHooksController) sync(ctx context.Context, _ factory.SyncContext) error {
+	// stop if the machine API is not functional
+	if isFunctional, err := c.machineAPIChecker.IsFunctional(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	// step 1: add hooks for voting members
+	if err := c.addDeletionHookToVotingMembers(ctx); err != nil {
```

Comment:
What is the disadvantage of simply adding the hook to every master machine?  that would effectively eliminate one class of races since the deletion hook would be present long before etcd was present on the machine.

Reply:
It won't solve the issue. In the case that a Machine is deleted before the member is promoted, the etcd operator is expected to not promote the new member, and remove the deletion hook to allow the Machine to be removed from the cluster.